mistletoe/markdown_renderer.py

```diff
@@ -0,0 +1,542 @@
+    """
+
+    repr_attributes = ("label", "dest", "title")
+
+    def __init__(self, match):
+        self.label, self.dest, self.title, self.dest_type, self.title_delimiter = match
+
+
+class LinkReferenceDefinitionBlock(block_token.Footnote):
+    """
+    A sequence of link reference definitions.
+    This is a container block token. Its children are link reference definition tokens.
+
+    This class inherits from Footnote and modifies the behavior of the constructor,
+    to keep the tokens in the AST.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        obj = object.__new__(cls)
+        obj.__init__(*args, **kwargs)
+        return obj
+
+    def __init__(self, matches):
+        self.children = list(map(LinkReferenceDefinition, matches))
+
+
+class Fragment:
+    """
+    Markdown fragment. Used when rendering trees of span tokens into flat sequences.
+    May carry additional data in addition to the text.
+
+    Attributes:
+        text (str): markdown fragment
+    """
+
+    def __init__(
+        self,
+        text: str,
+        **extras,
+    ):
+        self.text = text
+        for k, v in extras.items():
+            setattr(self, k, v)
+
+
+class MarkdownRenderer(BaseRenderer):
+    """
+    Markdown renderer.
+
+    Designed to make as "clean" a roundtrip as possible, markdown -> parsing -> rendering -> markdown,
+    except for nonessential whitespace. Except when rendering with word wrapping enabled.
+
+    Includes HTMLBlocks and HTMLSpans in the parsing.
+    """
+
+    _whitespace = re.compile(r"\s+")
+
+    def __init__(self, *extras):
+        block_token.remove_token(block_token.Footnote)
+        super().__init__(
+            *chain(
+                (
+                    block_token.HTMLBlock,
+                    span_token.HTMLSpan,
+                    BlankLine,
+                    LinkReferenceDefinitionBlock,
+                ),
+                extras,
+            )
+        )
+        self.render_map["SetextHeading"] = self.render_setext_heading
+        self.render_map["CodeFence"] = self.render_fenced_code_block
+        self.render_map[
+            "LinkReferenceDefinition"
+        ] = self.render_link_reference_definition
+
+    def render(self, token: token.Token, max_line_length: int = None) -> str:
+        """
+        Renders the tree of tokens rooted at the given token into markdown.
+
+        If `max_line_length` is specified, the document is word wrapped to the
+        specified line length. Otherwise the formatting from the original (parsed)
+        document is retained as much as possible.
+        """
+        if isinstance(token, block_token.BlockToken):
+            lines = self.render_map[token.__class__.__name__](
+                token, max_line_length=max_line_length
+            )
+        else:
+            lines = self.span_to_lines([token], max_line_length=max_line_length)
+
+        return "".join(map(lambda line: line + "\n", lines))
```

Comment:
OK, can you just explain a little bit why we add `\n` explicitly here, while in the `BaseRenderer`, there is just `''.join(map(self.render, token.children))`?

Reply:
Sure, I'll try!

Here in MarkdownRenderer.render() we can pass in either a block token or a span token for rendering. If it's a block token, we'll look it up in the render_map and find the appropriate rendering method to call. All rendering methods for block tokens return a sequence of lines (without trailing endlines). If, on the other hand, it's a span token, then we call the self.span_to_lines method to render it into a sequence of lines (also without trailing endlines). The string join operation smacks an endline on each line before the join.

Why then don't the rendering methods already include the endlines? That's really just a matter of taste. I think it's cleaner to work with lines without the endlines and apply the endlines in a single place, that's all.

In the BaseRenderer, *all* the token rendering methods (block and span) return plain strings to be joined. Hence the render method can be simpler.

One could argue that this design where the overridden methods in a child class returns a different type of data from the base class isn't exactly a schoolbook example of inheritance. But I think the pros are bigger than the cons in this case and that it's actually pretty neat. The block tokens behave very differently from the span tokens from the perspective of the Markdown renderer. These rendering methods aren't supposed to be called by anyone outside the class. And python, being an untyped language, has no problem with it.

Reply:
OK, thanks for the explanation. So I think we can leave as-is, at least for now, if it doesn't seem to break or "break" anything now or in the near future. :)